### DISCUSSION TITLE
`nu-complete` missing

### DISCUSSION THREAD

**Discussion by theAkito** [nushell/nushell#9989] [Q&A] [answered]
Maybe I missed the memo, but I could not find any blog post, changelog or whatever regarding `nu-complete` being removed.

One of my libraries stopped worked, because it says it cannot find stuff like `nu-complete dir`, `nu-complete file`, etc.

Is it deprecated or renamed or what happened? How do I fix my library?

**Comment by fdncred** [comment]
I thought `nu-complete` was just the name of a custom command.

Seems like the `nu-complete blah` tools in std lib still work.
use std
std help alias <tab>

**Reply by theAkito** [reply]
Thank you for your comment.

I'm sorry, I do not fully understand.

I tried `std help alias` & it worked, but `std help nu-complete` did not work.

Could you please explain?

**Reply by fdncred** [reply]
The way it's used in the std lib, nu-complete isn't a command, but `nu-complete list-aliases` is a command.
```nushell
def "nu-complete list-aliases" [] {
    scope aliases | select name usage | rename value description
}
```
but then it's uses for completions here
```nushell
export def aliases [
    ...alias: string@"nu-complete list-aliases"  # the name of alias to get help on
    --find (-f): string  # string to find in alias names
] {
```
which gets called when you do `std help aliases <tab>`

This `nu-complete` thing, as it's used here, is literally just a name. It could be changed to `bob list-aliases` and then used as `...alias: string@"bob list-aliases"`.

**Reply by theAkito** [reply]
Okay, I get it now.

I didn't see, that I actually created those completions myself. I thought, they are part of Nushell or something.

Now, the question is, why the original module cannot use those custom completions, when I imported the sub-module, which defined & exported those definitions.

They worked at some point, with an older Nu version, because I created & tested it myself back then.

Does it only work, when I `source` the module? Why, though?

I `use` it properly, so those completions should be available in the module calling these custom `nu-complete` commands.

**Reply by fdncred** [reply]
You should get on Discord and ask around. I don't use modules much. @amtoine is more familiar with them than I.

**Reply by amtoine** [reply, marked as answer]
`nu-complete foo` is just a convention i took from the book to write custom completions.
it's an internal implementation detail that gives completions to the `std help` commands, thus it's not `export`ed and not available to the outside of the `std` module :yum: 

i believe these completion commands could be called anything, as long as they are just next to the commands in which they are used :ok_hand:

**Reply by theAkito** [reply]
@amtoine

I `use` a module with custom completions, but they are apprently not imported into the `use`ing module, since the compilation fails with `undefined command`, when using them in `def`'s parameters, like `string@"nu-complete dir"`.

Custom completions in the same module, work though.

**Reply by amtoine** [reply]
yup that's what i meant by "next to the commands in which they are used" :yum: 

i think this is an annoying limitation tbh :thinking:

**Reply by theAkito** [reply]
Okay, I think it would make sense to add this fact to the official documentation.

Is it not possible to remove this limitation?

Why did it work before & does not work anymore?

**Reply by amtoine** [reply]
> Okay, I think it would make sense to add this fact to the official documentation.
> 
for the time being, to avoid confusion, it might be a great thing to add this to the documentation :+1: 

> Is it not possible to remove this limitation?
> 
i bet it is :yum: 
just need someone to implement the relaxation :smirk:

> Why did it work before & does not work anymore?
>
i do not remember this working in the past, maybe before i jumped in Nushell :yum:

**Reply by theAkito** [reply]
> i do not remember this working in the past, maybe before i jumped in Nushell 😋

Yes, I randomly found this in one of my modules, I have composed & entirely written myself. So, it must've worked in the past, else I wouldn't have been able to reproduce the expected behaviour.

Either way, I understand now what to do about my issue & hope this will be fixed for everyone at some point. With Nu, it should be natural have custom completions be importable via module.

If you give me some resource on how to add this to the documentation, I could do it myself.

**Reply by amtoine** [reply]
> > i do not remember this working in the past, maybe before i jumped in Nushell yum
> 
> Yes, I randomly found this in one of my modules, I have composed & entirely written myself. So, it must've worked in the past, else I wouldn't have been able to reproduce the expected behaviour.
> 
> Either way, I understand now what to do about my issue & hope this will be fixed for everyone at some point. With Nu, it should be natural have custom completions be importable via module.
> 
yup, could you file an issue about that if there is none? :innocent: 

> If you give me some resource on how to add this to the documentation, I could do it myself.
>
you'd need to change [this page](https://github.com/nushell/nushell.github.io/blob/main/book/custom_completions.md?plain=1#custom-completions) :+1: 

- fork the repo
- pull down your copy locally
- create a new branch off the `main` one
- apply some changes
- push it to your fork
- open a PR
is the general way to contribute :yum:
